test/router-luatest/router_test.lua

```diff
@@ -285,6 +285,190 @@ g.test_map_callrw_raw = function(g)
+    local bids2 = vtest.storage_get_n_buckets(g.replica_2_a, 1)
+
+    --
+    -- Successful partial map (single replicaset).
+    --
+    res = g.router:exec(function(bid1, bid2)
+        local val, err, err_uuid = ivshard.router.map_part_callrw(
+            {bid1, bid2},
+            'do_map',
+            {3},
+            {timeout = iwait_timeout})
+        return {
+            val = val,
+            err = err,
+            err_uuid = err_uuid,
+        }
+    end, {bids1[1], bids1[2]})
+    t.assert(not res.err)
+    t.assert(not res.err_uuid)
+    expected = {
+        [rs1_uuid] = {{1, 3}},
+    }
+    t.assert_equals(res.val, expected)
+    --
+    -- Successful partial map (multiple replicasets)
+    --
+    res = g.router:exec(function(bid1, bid2)
+        local val, err, err_uuid = ivshard.router.map_part_callrw({bid1, bid2}, 'do_map', {42},
+            {timeout = iwait_timeout})
+        return {
+            val = val,
+            err = err,
+            err_uuid = err_uuid,
+        }
+    end, {bids1[1], bids2[1]})
+    t.assert(not res.err)
+    t.assert(not res.err_uuid)
+    expected = {
+        [rs1_uuid] = {{1, 42}},
+        [rs2_uuid] = {{2, 42}},
+    }
+    t.assert_equals(res.val, expected)
+    --
+    -- Ref stage.
+    --
+    -- First move some buckets from rs1 to rs2 and then pause gc on rs1.
+    -- As a result, the buckets will be in the SENT state on rs1 and
+    -- in the ACTIVE state on rs2.
+    g.replica_1_a:exec(function(bid1, bid2, to)
+        ivshard.storage.internal.errinj.ERRINJ_BUCKET_GC_PAUSE = true
+        ivshard.storage.bucket_send(bid1, to)
+        ivshard.storage.bucket_send(bid2, to)
+    end, {bids1[1], bids1[2], rs2_uuid})
+    -- The buckets are ACTIVE on rs2, so the partial map should succeed.
+    res = g.router:exec(function(bid1, bid2)
+        local val, err, err_uuid = ivshard.router.map_part_callrw(
+            {bid1, bid2}, 'do_map', {42}, {timeout = iwait_timeout})
+        return {
+            val = val,
+            err = err,
+            err_uuid = err_uuid,
+        }
+    end, {bids1[1], bids1[2]})
+    t.assert(not res.err)
+    t.assert(not res.err_uuid)
+    expected = {
+        [rs2_uuid] = {{2, 42}},
+    }
+    t.assert_equals(res.val, expected)
+    -- But if we use some active bucket from rs1, the partial map should fail.
+    -- The reason is that the moved buckets are still in the SENT state and
+    -- we can't take a ref.
+    res = g.router:exec(function(bid1, bid2)
+        local val, err, err_uuid = ivshard.router.map_part_callrw(
+            {bid1, bid2}, 'do_map', {42}, {timeout = iwait_timeout})
+        return {
+            val = val,
+            err = err,
+            err_uuid = err_uuid,
+        }
+    end, {bids1[1], bids1[3]})
+    t.assert(res.val == nil)
+    t.assert(res.err)
+    t.assert_equals(res.err_uuid, rs1_uuid)
+    g.replica_1_a:exec(function()
+        ivshard.storage.internal.errinj.ERRINJ_BUCKET_GC_PAUSE = false
+    end)
```

Comment:
Please, add `_G.bucket_gc_wait()` after the errinj is unset. Otherwise the GC might have not enough time to drop the `SENT` buckets and an attempt to send them back would fail, making the test flaky.

Reply:
Fixed